### PR TITLE
Add Antithesis intrumentation

### PR DESCRIPTION
**Pull request by Bronek**
## High Level Overview of Change

Add Antithesis instrumentation.

### Context of Change

This change adds the support for [continuous testing (fuzzing)](https://youtu.be/KoxXb2mBVuY) of `rippled` on [Antithesis platform](https://antithesis.com/docs/introduction/how_antithesis_works.html). A copy of [Antithesis C++ SDK](https://github.com/antithesishq/antithesis-sdk-cpp/) is added to `external/antithesis-sdk` (with `clang-format` disabled, to make it easier to keep up).

The instrumentation macros `ASSERT` and `UNREACHABLE` are defined  inside `include/xrpl/beast/utility/instrumentation.h`. In a normal build, these macros are just wrappers (with added name) for regular `assert` macro. In such build, there is no dependency on `external/antithesis-sdk`. The documentation of macros is also in `include/xrpl/beast/utility/instrumentation.h`, and the naming convention for instrumentation macros has been added to `CONTRIBUTING.md`

If `voidstar` CMake option is enabled (which is only supported with clang compiler, version 16 or later, running on Linux, and requires a Debug build), these macros will register and forward instrumentation calls to `external/antithesis-sdk`; this in turn informs the Antithesis platform of all the contracts. In case if a contract is violated during fuzzing, a report will be created for `rippled` developers for troubleshooting. The purpose of name in instrumentation macros is to provide instrumentation calls with stable identity, which does not rely on line numbers.

The `voidstar` CMake option is not exposed in `conanfile.py` because it is not useful for programs with a dependency on this project; it is only useful when building `rippled` binary with the intent of running it on the Antithesis platform. This means that, for ordinary users, macros `ASSERT` or `UNREACHABLE` are just like named `assert`.

Both `ASSERT` or `UNREACHABLE` are so-called "always assertions" - that is, they are [test properties](https://antithesis.com/docs/introduction/how_antithesis_works.html#test-properties), which must be at all times true. Antithesis also supports a ["sometimes assertions"](https://antithesis.com/docs/best_practices/sometimes_assertions.html) which this PR does **not** define; we will add them later, on a case-by-case basis.

Also, this PR does not use ["rich assertions"](https://antithesis.com/blog/zelda/) which are added in [Antithesis SDK 0.4](https://github.com/antithesishq/antithesis-sdk-cpp/releases/tag/v0.4.0); this will be also used in a separate PR, in order to keep this PR as close as possible to regular `assert` model (even though this makes fuzzing slower to discover bugs, than it would have been otherwise).

#### Summary of changes

1. Define and document instrumentation macros in `include/xrpl/beast/utility/instrumentation.h`
2. Add external dependency `antithesis-sdk-cpp` in `external/antithesis-sdk`
3. Add support for `antithesis-sdk-cpp` in `CMakeLists.txt` and `cmake/RippledInstall.cmake`
4. Add new CMake option `voidstar` to `cmake/RippledCompiler.cmake`, `cmake/RippledCore.cmake` and `cmake/RippledSettings.cmake`
5. Set compilation option `--build-id` unconditionally in `cmake/RippledCompiler.cmake`
6. Update `CONTRIBUTING.md` to document contracts and instrumentation
7. Add `UNREACHABLE` in `LogicError` in `src/libxrpl/basics/contract.cpp`
8. All the remaining code changes (_several hundred lines_) are replacing `assert` with either `ASSERT` or `UNREACHABLE` (where suitable), with added name
9. The naming guidance is explained inside `CONTRIBUTING.md`

I strived to make the naming of contracts consistent everywhere, however that was difficult given the size of change.

#### Macro `ASSERT` is `ALWAYS_OR_UNREACHABLE` rather than `ALWAYS`

Instrumentation macros are registered in the Antithesis platform and they can be _also_ used for the purpose of coverage reporting (that is, whether or not a given contract has been reached is also a test property on the Antithesis platform). The semantics of `ALWAYS` macro (as defined in [Antithesis C++ SDK](https://antithesis.com/docs/using_antithesis/sdk/cpp/assert.html#always-assertions)) is not only to check that a condition is always true, it is **also** that the given line of code is _always_ executed during fuzzing. Since we do not expect all `assert`s to be always executed by a program during testing, we need a different macro, with the semantics that the condition is true **if the given line is executed**, otherwise the test platform should ignore the fact that a given line was **not executed**. This is the semantics of `ALWAYS_OR_UNREACHABLE` macro.

#### Macro `ASSERT` is not the same as `assert`

There are also locations where we do not use `ASSERT` or `UNREACHABLE` because the old style `assert` or `assert(false)` make more sense:
  * `constexpr` functions (`ASSERT` would fail compilation)
  * unit tests i.e. files under `src/test` (no need for instrumentation)
  * unit tests-related modules (files under `beast/test` and `beast/unit_test`; no need for instrumentation)

Also, as explained in `include/xrpl/beast/utility/instrumentation.h`, there are minor differences between `ASSERT` and `assert`:
 * `ASSERT` must have an unique name (naming convention in CONTRIBUTING.md)
 * the condition (which comes first) must be *implicitly* convertible to bool
 * during fuzzing, the program will continue execution past a failed `ASSERT` (like Release build) however the parameters will be fully evaluated (which is why CMake option `voidstar` requires Debug build) 

#### No change to the github workflow, yet

We need to upgrade clang compiler from version 14 to version no older than 16 in order to enable `voidstar` option, which would be needed to validate that the instrumentation macros do work with the the headers in `antithesis-sdk-cpp` (in particular that the `ASSERT` conditions are **implicitly** convertible to `bool`). I raised an issue https://github.com/XRPLF/rippled/issues/5153 for this upgrade. Once it is done, I intend to submit a separate PR which will define a new job to build with the `voidstar` option.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### CODE REVIEW

File: Builds/CMake/RippledCompiler.cmake
Comment (Bronek) [inline Builds/CMake/RippledCompiler.cmake:123]: Th unconditional addition of `--build-id` is intentional, but could be made conditional (on `voidstar`), like other options are. I just think it is a useful option to have even in a regular build.

**Comment by Bronek** [comment]
The choice of cmake option name `voidstar` comes from the reference to `libvoidstar.so` in the implementation of the Antithesis instrumentation - which is inside `external/antithesis-sdk/antithesis_instrumentation.h` file. I am open to replacing it with something else, e.g. `instrumentation`.

**Comment by Bronek** [comment]
TODO: Alter the consensus threshold to 80% (only when `voidstar` enabled),  so we can test the system with 6 validators only.

**Comment by intelliot** [comment]
6 validators is a good start and should yield useful results.

But just curious, how difficult would it be to later scale up the test to 35 (or more) validators?

(This would make the testing even more realistic and predict what could happen in the future if more validators are added)

File: include/xrpl/basics/instrumentation.h
Comment (thejohnfreeman) [inline include/xrpl/basics/instrumentation.h:29]: Can this just be `#error`? Maybe even just:

```cpp
#ifdef NDEBUG
#error "Antithesis instrumentation requires Debug build"
#endif
#include <antithesis_sdk.h>
```

File: include/xrpl/basics/instrumentation.h
Comment (thejohnfreeman) [inline include/xrpl/basics/instrumentation.h:33]: Wrap `cond` in parentheses?

File: Builds/levelization/results/loops.txt
Comment (thejohnfreeman) [inline Builds/levelization/results/loops.txt:9]: This is very disappointing to me. Wondering what you think of some alternatives.

a. We cannot just directly use `ALWAYS_OR_UNREACHABLE` in `xrpl/beast/` because it is not defined on non-Linux platforms.
b. We could write our own `<cassert>` and put it on the include path before the standard library.
c. We could not rewrite assertions within `xrpl/beast/`.
d. We could move the instrumentation header into `xrpl/beast/`.
e. We could add a new module, at a level beneath the `beast` module, that includes the instrumentation header.

**Review by thejohnfreeman — CHANGES_REQUESTED** [review]
Am I correct in guessing that all of the edits under `include/` and `src/` are just changing assertions to use the new macros?

**Comment by Bronek** [comment]
> 6 validators is a good start and should yield useful results.
> 
> But just curious, how difficult would it be to later scale up the test to 35 (or more) validators?
> 
> (This would make the testing even more realistic and predict what could happen in the future if more validators are added)

It's just a question of CPU utilisation on the Antithesis platform, that is balancing price/speed

File: include/xrpl/basics/instrumentation.h
Comment (Bronek) [reply include/xrpl/basics/instrumentation.h:29]: Yes, thanks will do that.

File: include/xrpl/basics/instrumentation.h
Comment (Bronek) [reply include/xrpl/basics/instrumentation.h:33]: yes, good point.

File: Builds/levelization/results/loops.txt
Comment (Bronek) [reply Builds/levelization/results/loops.txt:9]: I like d. Would that be ok for you as well ?

**Comment by Bronek** [comment]
> Am I correct in guessing that all of the edits under `include/` and `src/` are just changing assertions to use the new macros?

Yes.

**Review by thejohnfreeman — APPROVED** [review]

**Comment by Bronek** [comment]
Note, recent set of changes are related to the addition of names to instrumentation - which helps to ensure that Antithesis testing reports do not rapidly become obsolete.

**Comment by Bronek** [comment]
> > Am I correct in guessing that all of the edits under `include/` and `src/` are just changing assertions to use the new macros?
> 
> Yes.

@thejohnfreeman not anymore. With the commit  https://github.com/XRPLF/rippled/pull/5042/commits/8d7cd9c558aeea3eabe490342a926985b10c552b all the asserts have now gained names, which I tried to deduce by reading the surrounding code. This was needed to ensure that Antithesis reports do not lose the history of passing/failing asserts, as the location of asserts in code changes due to code churn.

File: include/xrpl/beast/utility/instrumentation.h
Comment (bthomee) [inline include/xrpl/beast/utility/instrumentation.h:44]: The `ALWAYS` and `REACHABLE` have the same usage description, but must have different definitions. Can you clarify their differences here?

File: include/xrpl/beast/utility/instrumentation.h
Comment (bthomee) [inline include/xrpl/beast/utility/instrumentation.h:38]: You can avoid guarding the `#include <antithesis_sdk.h>` as it already contains these define statements if their SDK is not enabled.

You could do something like:

```
#ifndef ENABLE_VOIDSTAR
#define NO_ANTITHESIS_SDK
#endif
#include <antithesis_sdk.h>
```

I left off the `NDEBUG` check above, because the cmake config already takes care of when the voidstar option can and cannot be turned on. And the Antithesis header then has the following check to determine whether or not to enable the SDK: `#if defined(NO_ANTITHESIS_SDK) || __cplusplus < 202000L || (defined(__clang__) && __clang_major__ < 16)`.

That being said, the Antithesis header does have a few system library includes plus the namespace definition containing a struct (top of file) and template function (bottom of file). If it is indeed a concern that these would be included in all builds, including releases, then your current code would be fine. That being said, if you're open to modifying the `antithesis_sdk.h` file then you could possibly move that `#if defined(NO_ANTITHESIS_SDK) ...` statement around the namespace or at least the struct/function to get the best of both worlds.

File: include/xrpl/beast/utility/instrumentation.h
Comment (Bronek) [reply include/xrpl/beast/utility/instrumentation.h:44]: the difference is that one takes `cond` parameter and the other does not - I think this is visible enough ?

File: include/xrpl/beast/utility/instrumentation.h
Comment (Bronek) [reply include/xrpl/beast/utility/instrumentation.h:38]: The idea was to entirely exclude the dependency on `external/antithesis-sdk` if the build is not using the `voidstar` option. The reason is that, at least for the time being, we want `rippled` to build successfully on a wider selection of compilers than supported by `external/antithesis-sdk` - which includes clang 15, which would trip on this:

```
#if defined(__clang__) && __clang_major__ < 16
    #error "The Antithesis C++ API requires clang version 16 or higher"
#endif
```

Having said that I like the idea of not having to repeat definitions which are in this section of `#include <antithesis_sdk.h>` ; maybe I can talk Antithesis into doing something to make this header easier to swallow for clang 15 (when `NO_ANTITHESIS_SDK` is defined)

Comment (Bronek) [reply include/xrpl/beast/utility/instrumentation.h:38]: No wait, I would still need to guard `#include <antithesis_sdk.h>`, because (if `voidstar` is not being used) I need to define all the polyfills for instrumentation macros. The polyfills from `<antithesis_sdk.h>` are not good, as they do not `assert`.

Comment (bthomee) [reply include/xrpl/beast/utility/instrumentation.h:38]: Ah yes, you are correct - they just define the names but don't attach any actions to them.

File: include/xrpl/beast/utility/instrumentation.h
Comment (bthomee) [reply include/xrpl/beast/utility/instrumentation.h:44]: It is visible, but requires the reader to look at both the define statements and the comments to figure that out. I figured the reachable description could be something like `REACHABLE if it must be reached during fuzzing provided the specified condition holds`.

Comment (Bronek) [reply include/xrpl/beast/utility/instrumentation.h:44]: It's not "provided" - this would imply that, if the condition does not hold, the location may not be reached. I will figure out better explanation.

File: src/libxrpl/basics/contract.cpp
Comment (bthomee) [inline src/libxrpl/basics/contract.cpp:52]:
```suggestion
    UNREACHABLE("ripple::detail::LogicError", {{"message", s}});
```

File: src/libxrpl/json/Object.cpp
Comment (bthomee) [inline src/libxrpl/json/Object.cpp:171]:
```suggestion
    UNREACHABLE("Json::Array::append : invalid type"); 
```

The `UNREACHABLE` statement now makes it clear no path can get here.

File: src/libxrpl/json/Object.cpp
Comment (bthomee) [inline src/libxrpl/json/Object.cpp:206]:
```suggestion
    UNREACHABLE("Json::Object::set : invalid type");
```

File: src/libxrpl/json/json_reader.cpp
Comment (bthomee) [inline src/libxrpl/json/json_reader.cpp:956]:
```suggestion
```

Commented-out code can be removed. Moreover, if you want to keep it then the `ASSERT` macro must be fully implemented, including giving it a name.

File: src/test/csf/Histogram.h
Comment (bthomee) [inline src/test/csf/Histogram.h:22]:
```suggestion
```

Include does not seem to be used, so can be removed.

I see this occurring in several files. I gave a few additional suggestions for removal below, but please check all files to see if the include isn't unnecessarily present if none of the macros are called

File: src/test/jtx/impl/amount.cpp
Comment (bthomee) [inline src/test/jtx/impl/amount.cpp:23]:
```suggestion
```

File: src/test/overlay/short_read_test.cpp
Comment (bthomee) [inline src/test/overlay/short_read_test.cpp:24]:
```suggestion
```

File: src/test/unit_test/multi_runner.cpp
Comment (bthomee) [inline src/test/unit_test/multi_runner.cpp:23]:
```suggestion
```

File: src/xrpld/app/consensus/RCLConsensus.cpp
Comment (bthomee) [inline src/xrpld/app/consensus/RCLConsensus.cpp:95]: Shouldn't this file have a `#include <xrpl/beast/utility/instrumentation.h>` statement? I presume it compiles due to transitive includes, but it would be better for the file to fully mention all dependencies as imports.

I see this across many other files too, so please double-check all.

File: src/xrpld/app/tx/detail/InvariantCheck.cpp
Comment (bthomee) [inline src/xrpld/app/tx/detail/InvariantCheck.cpp:383]: Would this be a good candidate for `SOMETIMES`? If the fuzzer never can get the enforce variable to be true, perhaps it indeed is unused.

File: src/xrpld/app/tx/detail/OfferStream.cpp
Comment (bthomee) [inline src/xrpld/app/tx/detail/OfferStream.cpp:345]:
```suggestion
                "vs XRP offer");
```

The extra comment is now superfluous.

File: src/xrpld/app/tx/detail/SetSignerList.cpp
Comment (bthomee) [inline src/xrpld/app/tx/detail/SetSignerList.cpp:134]:
```suggestion
        "ripple::SetSignerList::doApply : invalid operation");
```

File: src/xrpld/core/JobTypes.h
Comment (bthomee) [inline src/xrpld/core/JobTypes.h:69]: As before, would this be a candidate for using `SOMETIMES`, if you don't know if it's actually used?

File: external/antithesis-sdk/antithesis_instrumentation.h
Comment (Bronek) [reply external/antithesis-sdk/antithesis_instrumentation.h:102]: Nope, I'd rather take `external/antithesis-sdk`  exactly as-is, so we do not have think hard when upgrading to a newer version

File: external/antithesis-sdk/antithesis_instrumentation.h
Comment (Bronek) [reply external/antithesis-sdk/antithesis_instrumentation.h:110]: I'd rather use `external/antithesis-sdk` exactly as provided

File: src/xrpld/core/JobTypes.h
Comment (Bronek) [reply src/xrpld/core/JobTypes.h:69]: No, `SOMETIMES` is only a hint for the fuzzer, it does not check anything.

File: src/xrpld/app/tx/detail/InvariantCheck.cpp
Comment (Bronek) [reply src/xrpld/app/tx/detail/InvariantCheck.cpp:383]: The `[[maybe_unused]]` attribute means that the compiler should not warn when the variable is never used. This variable `enforce` is only ever used inside an `ASSERT` (down below in this function), meaning the compiler will see it being used in any Debug build (including `voidstar` one) but will _not_ see it being used in a Release build. That's how we use `[[maybe_unused]]` in several other location in this project. On the other hand `SOMETIMES` disables check altogether, and that's not what we want - since before this PR we had actual `assert` in this location.

File: src/xrpld/app/consensus/RCLConsensus.cpp
Comment (Bronek) [reply src/xrpld/app/consensus/RCLConsensus.cpp:95]: great point.

File: src/xrpld/app/paths/detail/StrandFlow.h
Comment (Bronek) [reply src/xrpld/app/paths/detail/StrandFlow.h:850]: there was commented-out `assert(0)` before; replacing it with an actual `UNREACHABLE()` would change the program behaviour from "we put a comment here" to "we crash", and that's not a good PR for this kind of change.

File: CONTRIBUTING.md
Comment (Bronek) [reply CONTRIBUTING.md:371]: We do not seem to care that much about 80 chars per line in `.md` files, so :man_shrugging:

File: external/antithesis-sdk/antithesis_instrumentation.h
Comment (Bronek) [reply external/antithesis-sdk/antithesis_instrumentation.h:94]: I leave that to the authors of Antithesis SDK ; they decided that `antithesis_load_libvoidstar` should have no parameters and that seems OK to me. If it were to call `trace_pc_guard_init(start, stop);` then it would have to also receive `start` and `stop` parameters.

File: src/libxrpl/basics/contract.cpp
Comment (Bronek) [reply src/libxrpl/basics/contract.cpp:52]: I'd rather keep this one; it is special because there are different ways `LogicError` may happen, which is also why the extra parameter.

File: src/xrpld/app/paths/detail/StrandFlow.h
Comment (bthomee) [reply src/xrpld/app/paths/detail/StrandFlow.h]: Let me rephrase: It was a commented-out `assert`, and you've replaced it by a commented-out `UNREACHABLE`. My assumption here is that you want to leave this here commented out, so it can be re-enabled in the future when needed. If so, the old `assert(0)` statement would compile fine but the new `UNREACHABLE()` statement would fail to compile as that macro expects at least a name to be provided.

**Review by bthomee — APPROVED** [review]

**Review by bthomee — APPROVED** [review]

**Comment by Bronek** [comment]
Blocked on #5143 and #5103

**Review by intelliot — APPROVED** [review]
note: this will be merged _after_ rippled 2.3.0

**Comment by ximinez** [comment]
@Bronek Before I merge this, can you write a commit message with a consise summary?

**Comment by Bronek** [comment]
> @Bronek Before I merge this, can you write a commit message with a consise summary?

Add Antithesis instrumentation

* Copy Antithesis SDK version 0.4.0 to directory external/
* Add build option `voidstar` to enable instrumentation with Antithesis SDK
* Define instrumentation macros ASSERT and UNREACHABLE in terms of regular C assert
* Replace all asserts with named ASSERT or UNREACHABLE
* Add UNREACHABLE to LogicError
* Document instrumentation macros in CONTRIBUTING.md

**Comment by ximinez** [comment]
Merged manually as d7e949193fe3a42c67f36c021daba9fd0d52780c.
```
$ git diff --exit-code Bronek/feature/antithesis-instrumentation && echo Identical
Identical
```

**Pull request closed without merging**
